src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/DefaultRazorLanguageClientMiddleLayer.cs

```diff
@@ -0,0 +1,215 @@
+        {
+            return null;
+        }
+
+        public override async Task<JToken> HandleRequestAsync(string methodName, JToken methodParam, Func<JToken, Task<JToken>> sendRequest)
+        {
+            if (methodName == Methods.TextDocumentOnTypeFormattingName)
+            {
+                var emptyResult = JToken.FromObject(Array.Empty<TextEdit>());
+                var requestParams = methodParam.ToObject<DocumentOnTypeFormattingParams>();
+                if (requestParams.Options.OtherOptions == null)
+                {
+                    requestParams.Options.OtherOptions = new Dictionary<string, object>();
+                }
+
+                requestParams.Options.OtherOptions[LanguageServerConstants.ExpectsCursorPlaceholderKey] = true;
+                var token = JToken.FromObject(requestParams);
+                var result = await sendRequest(token);
+                var edits = result?.ToObject<TextEdit[]>();
+                if (edits == null)
+                {
+                    return emptyResult;
+                }
+
+                await _joinableTaskFactory.SwitchToMainThreadAsync();
+
+                if (!_documentManager.TryGetDocument(requestParams.TextDocument.Uri, out var documentSnapshot))
+                {
+                    return emptyResult;
+                }
+
+                Position cursorPosition = null;
+                var filteredEdits = new List<TextEdit>(edits.Length);
+                foreach (var edit in edits)
+                {
+                    if (edit.NewText.Contains(LanguageServerConstants.CursorPlaceholderString))
+                    {
+                        var newEdit = new TextEdit();
+                        newEdit.Range = edit.Range;
+                        newEdit.NewText = edit.NewText.Replace(LanguageServerConstants.CursorPlaceholderString, string.Empty);
+                        filteredEdits.Add(newEdit);
+
+                        cursorPosition = edit.Range.Start;
+                    }
+                    else
+                    {
+                        filteredEdits.Add(edit);
+                    }
+                }
+
+                ApplyTextEdits(filteredEdits, documentSnapshot.Snapshot, documentSnapshot.Snapshot.TextBuffer);
+
+                if (cursorPosition != null)
+                {
+                    var fullPath = GetLocalFilePath(requestParams.TextDocument.Uri);
+                    VsShellUtilities.OpenDocument(_serviceProvider, fullPath, VSConstants.LOGVIEWID.TextView_guid, out _, out _, out _, out var textView);
+                    MoveCaretToPosition(textView, cursorPosition);
+                }
+
+                return emptyResult;
+            }
+            else
+            {
+                return await sendRequest(methodParam);
+            }
+        }
+
+        internal static void ApplyTextEdits(IEnumerable<TextEdit> textEdits, ITextSnapshot snapshot, ITextBuffer textBuffer)
```

Comment:
Everything below this came from [here](https://devdiv.visualstudio.com/DevDiv/_git/VSLanguageServerClient?path=%2Fsrc%2Fproduct%2FRemoteLanguage%2FImpl%2FUtilities.cs)

Reply:
Should probably move to someplace common.

Reply:
Should follow up with LSP client folks on their opinion here too

Reply:
> Should follow up with LSP client folks on their opinion here too

Do you mean if they'll expose it for us to use?

Reply:
Just got out of a meeting with editor folk. NM we can live with this as is until they provide snippet support.